core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/ast/SyntaxTreeBuilderTokenManager.java

```diff
@@ -3340,6 +3340,25 @@ public void SwitchTo(int lexState) {
 		}
 	}
 
+	public String backSlashCorrection(String image) {
+		StringBuilder newimg = new StringBuilder();
+		int imglen = image.length(), cnt = 0;
+		for (int i = 0; i < imglen; i++) {
+			if (image.charAt(i) == '\\') {
+				cnt++;
+			} else if (cnt % 2 == 1) {
+				newimg.append('\\');
+				cnt = 0;
+			}
+			newimg.append(image.charAt(i));
+		}
+		if (cnt % 2 == 1) {
+			newimg.append('\\');
+		}
+		String newImage = new String(newimg);
+		return newImage;
+	}
```

Comment:
Unfortunately, this is the wrong place to fix the issue. The SyntaxTreeBuilderTokenManager is generated code, created/updated when we run [JavaCC](https://javacc.github.io/javacc/) on the sparql.jjt file. Next time we regenerate the parser from the jjt file, your changes would be overwritten. 

It's been a while since I last used JavaCC, but I think the way to fix this would be by adding a new method to the SyntaxTreeBuilder class definition you will find at the top of the sparql.jjt file, a method that takes a regular expression token and corrects the escaping, and then make sure in the actual grammar production rule for the regex (lines 1619-1623) use that method to produce the correct token. Have a look at why we use a similar technique to create a datablock for INSERT DATA updates, using the `readToMatchingBrace` method. 

By the way there is an Eclipse plugin for JavaCC that makes working with it a lot easier: https://marketplace.eclipse.org/content/javacc-eclipse-plug